### PR TITLE
[CUSPARSE] Add sddmm! and gemvi! routines

### PR DESCRIPTION
**Pull request by amontoison**
- Interface `sddmm!` and `gemvi!`
- export functions added in #1611
- update the interface for mul! and ldiv!

### CODE REVIEW

**Comment by amontoison** [comment]
We need to define a function for `CuMatrix * CuSparseVector` because the `mul!` is not enough in that case.
https://github.com/JuliaSparse/SparseArrays.jl/blob/main/src/sparsevector.jl#L1736-L1743

File: lib/cusparse/generic.jl [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # generic APIs
 
-export vv!, sv!, sm!, gemm, gemm!
+export sparsetodense, densetosparse, gather!, scatter!, axpby!, rot!
```
Comment (maleadt) [inline lib/cusparse/generic.jl]: If we export this, I'd use `sparse_to_dense` and `dense_to_sparse`. Or maybe `to_dense` or even `dense` if we can disambiguate it sufficiently based on the arguments? Although that may be confusing for `sparse` (given that there already is a function with specific meaning).

File: lib/cusparse/generic.jl [hunk 2]
```diff
@@ -216,6 +217,11 @@ function mm!(transa::SparseChar, transb::SparseChar, alpha::Number, A::Union{CuS
         return out[]
     end
     with_workspace(bufferSize) do buffer
+        # Uncomment is we find a way to reuse the buffer
```
Comment (maleadt) [inline lib/cusparse/generic.jl]: Typo. Also, could you elaborate a little more, and/or link to an issue? Is https://github.com/JuliaGPU/CUDA.jl/issues/1362 related?

**Review by maleadt — CHANGES_REQUESTED** [review]
CI failure looks related.

File: lib/cusparse/generic.jl (hunk 1 above)
Comment (amontoison) [reply lib/cusparse/generic.jl]: You're right, we don't need to export `sparsetodense`, `densetosparse`.
We use `CuMatrix(A)`, `CuSparseMatriCSC(A)`, `CuSparseMatriCSR(A)`, `CuSparseMatriCOO(A)` in practice.
I will only export the other functions.

File: lib/cusparse/generic.jl (hunk 2 above)
Comment (amontoison) [reply lib/cusparse/generic.jl]: Yes, it's exactly the issue #1362!
I didn't know that @frapac already opened an issue about it.

**Review by maleadt — APPROVED** [review]

**Comment by amontoison** [comment]
I have an idea for #1403 (support dense matrix - sparse matrix products).
We could use sparse matrix - dense matrix products (`mm!` function).
I have some issues during the compilation (`TypeB` not defined...) and I don't know how to permute efficiently a CuMatrix or if we should return a `Transpose(CuMatrix)` / `Adjoint(CuMatrix)`.
In all cases, I will not add it in this PR.

```julia
tag_wrappers = ((identity, identity),
                (T -> :(HermOrSym{T, <:$T}), A -> :(parent($A))))
op_wrappers = (
    (identity, T -> 'N', identity),
    (T -> :(Transpose{<:T, <:$T}), T -> 'T', A -> :(parent($A))),
    (T -> :(Adjoint{<:T, <:$T}), T -> T <: Real ? 'T' : 'C', A -> :(parent($A)))
)

for (taga, untaga) in tag_wrappers, (wrapa, transa, unwrapa) in op_wrappers
    TypeA = wrapa(taga(:(DenseCuMatrix{T})))
    @eval begin
        for (tagb, untagb) in tag_wrappers, (wrapb, transb, unwrapb) in op_wrappers
            TypeB = wrapb(tagb(:(CuSparseMatrix{T})))

            @eval begin
                # Use the relations AB = (BᵀAᵀ)ᵀ and AB = (BᴴAᴴ)ᴴ to support dense matrix - sparse matrix products
                function LinearAlgebra.:(*)(A::$TypeA, B::$TypeB) where {T <: Union{Float16, ComplexF16, BlasFloat}}
                    mA, nA = size(A)
                    mB, nB = size(B)
                    C = CuMatrix{T}(undef, nB, mA)
                    transa = $transa(T)
                    transb = $transb(T)
                    (transa == 'N' && transb == 'N') && (transa2 = 'T' ; transb2 = 'T')
                    (transa == 'T' && transb == 'N') && (transa2 = 'N' ; transb2 = 'T')
                    (transa == 'C' && transb == 'N') && (transa2 = 'N' ; transb2 = 'C')
                    (transa == 'N' && transb == 'T') && (transa2 = 'T' ; transb2 = 'N')
                    (transa == 'N' && transb == 'C') && (transa2 = 'C' ; transb2 = 'N')
                    (transa == 'T' && transb == 'T') && (transa2 = 'N' ; transb2 = 'N')
                    (transa == 'C' && transb == 'C') && (transa2 = 'N' ; transb2 = 'N')
                    (transa == 'T' && transb == 'C') && error("dense matrix - sparse matrix product is not supported with transa == 'T' and transb == 'C'")
                    (transa == 'C' && transb == 'T') && error("dense matrix - sparse matrix product is not supported with transa == 'C' and transb == 'T'")
                    mm_wrapper(transb2, transa2, one(T), $(untagb(unwrapb(:B))), $(untaga(unwrapa(:A))), zero(T), C)
                    (transa == 'C' || transa == 'C') && (conj!(C))
                    permutedims(C)
                end
            end
        end
    end
end
```

**Pull request merged**